{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already up-to-date: google-cloud-storage in /opt/conda/lib/python3.7/site-packages (1.32.0)\n",
      "Requirement already satisfied, skipping upgrade: google-resumable-media<2.0dev,>=1.0.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage) (1.1.0)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-core<2.0dev,>=1.4.1 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage) (1.4.3)\n",
      "Requirement already satisfied, skipping upgrade: google-auth<2.0dev,>=1.11.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage) (1.22.1)\n",
      "Requirement already satisfied, skipping upgrade: requests<3.0.0dev,>=2.18.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-storage) (2.24.0)\n",
      "Requirement already satisfied, skipping upgrade: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /opt/conda/lib/python3.7/site-packages (from google-resumable-media<2.0dev,>=1.0.0->google-cloud-storage) (1.0.0)\n",
      "Requirement already satisfied, skipping upgrade: six in /opt/conda/lib/python3.7/site-packages (from google-resumable-media<2.0dev,>=1.0.0->google-cloud-storage) (1.15.0)\n",
      "Requirement already satisfied, skipping upgrade: google-api-core<2.0.0dev,>=1.19.0 in /opt/conda/lib/python3.7/site-packages (from google-cloud-core<2.0dev,>=1.4.1->google-cloud-storage) (1.22.4)\n",
      "Requirement already satisfied, skipping upgrade: rsa<5,>=3.1.4; python_version >= \"3.5\" in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.11.0->google-cloud-storage) (4.6)\n",
      "Requirement already satisfied, skipping upgrade: setuptools>=40.3.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.11.0->google-cloud-storage) (50.3.0)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.11.0->google-cloud-storage) (0.2.8)\n",
      "Requirement already satisfied, skipping upgrade: cachetools<5.0,>=2.0.0 in /opt/conda/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.11.0->google-cloud-storage) (4.1.1)\n",
      "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage) (2020.6.20)\n",
      "Requirement already satisfied, skipping upgrade: idna<3,>=2.5 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage) (2.10)\n",
      "Requirement already satisfied, skipping upgrade: chardet<4,>=3.0.2 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage) (3.0.4)\n",
      "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/conda/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-storage) (1.25.10)\n",
      "Requirement already satisfied, skipping upgrade: cffi>=1.0.0 in /opt/conda/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<2.0dev,>=1.0.0->google-cloud-storage) (1.14.3)\n",
      "Requirement already satisfied, skipping upgrade: googleapis-common-protos<2.0dev,>=1.6.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.19.0->google-cloud-core<2.0dev,>=1.4.1->google-cloud-storage) (1.52.0)\n",
      "Requirement already satisfied, skipping upgrade: protobuf>=3.12.0 in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.19.0->google-cloud-core<2.0dev,>=1.4.1->google-cloud-storage) (3.13.0)\n",
      "Requirement already satisfied, skipping upgrade: pytz in /opt/conda/lib/python3.7/site-packages (from google-api-core<2.0.0dev,>=1.19.0->google-cloud-core<2.0dev,>=1.4.1->google-cloud-storage) (2020.1)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1>=0.1.3 in /opt/conda/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.5\"->google-auth<2.0dev,>=1.11.0->google-cloud-storage) (0.4.8)\n",
      "Requirement already satisfied, skipping upgrade: pycparser in /opt/conda/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<2.0dev,>=1.0.0->google-cloud-storage) (2.20)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "#install upgraded google-cloud-storage\n",
    "#pip install --upgrade google-cloud-storage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Access Google Cloud Storage (GCS) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Access data files for H23 and parental test samples (first batch)\n",
    "\n",
    "from google.cloud import storage\n",
    "#google apis documentation: https://googleapis.dev/python/storage/latest/index.html\n",
    "bucket_name = 'sbi-sheffels-e'\n",
    "#prefix = 'runs/20210113-results/'\n",
    "\n",
    "storage_client = storage.Client() \n",
    "bucket = storage_client.get_bucket(bucket_name)\n",
    "#Get list of files in data/pE2.N54.fastq.00 folder of bucket\n",
    "#files = list(storage_client.list_blobs(bucket, prefix=prefix)) #blob = GCS name for file/object\n",
    "#for file in files:\n",
    " #   print(file) #print \"blob\" names\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringtie_file = \"runs/20210113-results/stringtieFPKM/ballgown/E2-78710_S1_L004_R1_001Aligned.sortedByCoord.out_ballgown/t_data.ctab\"\n",
    "stringtie_filename = \"t_data.ctab\"\n",
    "featureCount_file = \"runs/20210113-results/featureCounts/gene_counts/E2-78710_S1_L004_R1_001Aligned.sortedByCoord.out_gene.featureCounts.txt\"\n",
    "featureCount_filename = \"78710_featureCount.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get files\n",
    "blob_stringtie = storage.blob.Blob(stringtie_file,bucket)\n",
    "#download blob to AI Platform Notebook instance\n",
    "with open(stringtie_filename, \"wb\") as file_obj:\n",
    "    blob_stringtie.download_to_file(file_obj)\n",
    "\n",
    "blob_featureCount = storage.blob.Blob(featureCount_file,bucket)\n",
    "#download blob to AI Platform Notebook instance\n",
    "with open(featureCount_filename, \"wb\") as file_obj:\n",
    "    blob_featureCount.download_to_file(file_obj)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from statistics import mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCount = pd.read_csv(\"78710_featureCount.txt\", sep=\"\\t\", header = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Geneid', 'Chr', 'Start', 'End', 'Strand', 'Length', 'gene_name',\n",
       "       'E2-78710_S1_L004_R1_001Aligned.sortedByCoord.out.bam'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featureCount.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCount[\"Counts\"]=featureCount[\"E2-78710_S1_L004_R1_001Aligned.sortedByCoord.out.bam\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCount.drop(axis=1, columns=['Chr', 'Start', 'End', 'Strand', 'Length',\n",
    "       'E2-78710_S1_L004_R1_001Aligned.sortedByCoord.out.bam'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63672</th>\n",
       "      <td>ENSG00000224240</td>\n",
       "      <td>CYCSP49</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63673</th>\n",
       "      <td>ENSG00000227629</td>\n",
       "      <td>SLC25A15P1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63674</th>\n",
       "      <td>ENSG00000237917</td>\n",
       "      <td>PARP4P1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63675</th>\n",
       "      <td>ENSG00000231514</td>\n",
       "      <td>FAM58CP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63676</th>\n",
       "      <td>ENSG00000235857</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63677 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Geneid   gene_name  Counts\n",
       "0      ENSG00000223972     DDX11L1       0\n",
       "1      ENSG00000227232      WASH7P     276\n",
       "2      ENSG00000243485  MIR1302-10       0\n",
       "3      ENSG00000237613     FAM138A       0\n",
       "4      ENSG00000268020      OR4G4P       0\n",
       "...                ...         ...     ...\n",
       "63672  ENSG00000224240     CYCSP49       0\n",
       "63673  ENSG00000227629  SLC25A15P1       0\n",
       "63674  ENSG00000237917     PARP4P1       0\n",
       "63675  ENSG00000231514     FAM58CP       0\n",
       "63676  ENSG00000235857     CTBP2P1       0\n",
       "\n",
       "[63677 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featureCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCount.sort_values(\"gene_name\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13481</th>\n",
       "      <td>ENSG00000081760</td>\n",
       "      <td>AACS</td>\n",
       "      <td>2216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Geneid gene_name  Counts\n",
       "13481  ENSG00000081760      AACS    2216"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "featureCount.loc[featureCount[\"gene_name\"]==\"AACS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "featureCount[\"normalized_counts\"]=featureCount[\"Counts\"]/2216"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringtieCount = pd.read_csv(\"stringtie_counts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringtieCount.rename({\"Unnamed: 0\": \"Gene_name\"}, axis =1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare = pd.DataFrame()\n",
    "compare[\"Gene\"]=stringtieCount[\"Gene_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>abundance</th>\n",
       "      <th>counts</th>\n",
       "      <th>length</th>\n",
       "      <th>countsFromAbundance</th>\n",
       "      <th>counts_normalized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>197.741935</td>\n",
       "      <td>no</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7SK</td>\n",
       "      <td>0.184337</td>\n",
       "      <td>60.100913</td>\n",
       "      <td>2775.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>0.011352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>3.835476</td>\n",
       "      <td>226.055937</td>\n",
       "      <td>501.639247</td>\n",
       "      <td>no</td>\n",
       "      <td>0.042698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>1.681708</td>\n",
       "      <td>235.714062</td>\n",
       "      <td>1192.970295</td>\n",
       "      <td>no</td>\n",
       "      <td>0.044522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2450.400000</td>\n",
       "      <td>no</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56631</th>\n",
       "      <td>ZYG11AP1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>837.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56632</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>6.432696</td>\n",
       "      <td>6022.389568</td>\n",
       "      <td>7968.389620</td>\n",
       "      <td>no</td>\n",
       "      <td>1.137518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56633</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>130.485857</td>\n",
       "      <td>32288.723414</td>\n",
       "      <td>2106.115553</td>\n",
       "      <td>no</td>\n",
       "      <td>6.098740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56634</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>5.771518</td>\n",
       "      <td>7329.780256</td>\n",
       "      <td>10809.252789</td>\n",
       "      <td>no</td>\n",
       "      <td>1.384459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56635</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>13.912986</td>\n",
       "      <td>6440.800112</td>\n",
       "      <td>3940.163521</td>\n",
       "      <td>no</td>\n",
       "      <td>1.216548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56636 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene_name   abundance        counts        length countsFromAbundance  \\\n",
       "0       5S_rRNA    0.000000      0.000000    197.741935                  no   \n",
       "1           7SK    0.184337     60.100913   2775.000000                  no   \n",
       "2          A1BG    3.835476    226.055937    501.639247                  no   \n",
       "3      A1BG-AS1    1.681708    235.714062   1192.970295                  no   \n",
       "4          A1CF    0.000000      0.000000   2450.400000                  no   \n",
       "...         ...         ...           ...           ...                 ...   \n",
       "56631  ZYG11AP1    0.000000      0.000000    837.000000                  no   \n",
       "56632    ZYG11B    6.432696   6022.389568   7968.389620                  no   \n",
       "56633       ZYX  130.485857  32288.723414   2106.115553                  no   \n",
       "56634     ZZEF1    5.771518   7329.780256  10809.252789                  no   \n",
       "56635      ZZZ3   13.912986   6440.800112   3940.163521                  no   \n",
       "\n",
       "       counts_normalized  \n",
       "0               0.000000  \n",
       "1               0.011352  \n",
       "2               0.042698  \n",
       "3               0.044522  \n",
       "4               0.000000  \n",
       "...                  ...  \n",
       "56631           0.000000  \n",
       "56632           1.137518  \n",
       "56633           6.098740  \n",
       "56634           1.384459  \n",
       "56635           1.216548  \n",
       "\n",
       "[56636 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stringtieCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare[\"Stringtie\"]=stringtieCount[\"counts_normalized\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Can only compare identically-labeled Series objects",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-8a3583058f6a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcompare\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"featureCount\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfeatureCount\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mfeatureCount\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"gene_name\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mcompare\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Gene\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/pandas/core/ops/common.py\u001b[0m in \u001b[0;36mnew_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     63\u001b[0m         \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 65\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     66\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/pandas/core/arraylike.py\u001b[0m in \u001b[0;36m__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__eq__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__eq__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cmp_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moperator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__ne__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   4968\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4969\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_indexed_same\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4970\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Can only compare identically-labeled Series objects\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4971\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4972\u001b[0m         \u001b[0mlvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mextract_numpy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Can only compare identically-labeled Series objects"
     ]
    }
   ],
   "source": [
    "compare[\"featureCount\"]=featureCount.loc[featureCount[\"gene_name\"]==compare[\"Gene\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene_list = compare[\"Gene\"].tolist()\n",
    "non_zero_list = []\n",
    "value_list = []\n",
    "def make_avg_value(row):\n",
    "    gene = row[\"Gene\"]\n",
    "    value_list = []\n",
    "    non_zero_list = []\n",
    "    value_list = featureCount.loc[featureCount[\"gene_name\"]==gene][\"normalized_counts\"].tolist()\n",
    "    for value in value_list:\n",
    "        if value > 0:\n",
    "            non_zero_list.append(value)\n",
    "    if len(non_zero_list) == 0:\n",
    "        gene_average = 0\n",
    "    elif len(non_zero_list) == 1:\n",
    "        gene_average = non_zero_list[0]\n",
    "    elif len(non_zero_list) > 1:\n",
    "        gene_average = mean(non_zero_list)\n",
    "    return gene_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_avg_value(row):\n",
    "    gene = row[\"Gene\"]\n",
    "    value_list = featureCount.loc[featureCount[\"gene_name\"]==gene][\"normalized_counts\"].tolist()\n",
    "    gene_average = mean(value_list)\n",
    "    return gene_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "compare[\"featureCount_with0\"] = compare.apply(lambda row: make_avg_value(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Stringtie</th>\n",
       "      <th>featureCount</th>\n",
       "      <th>featureCount_with0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14799</th>\n",
       "      <td>EEF1A1</td>\n",
       "      <td>481.814875</td>\n",
       "      <td>401.810469</td>\n",
       "      <td>401.810469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14833</th>\n",
       "      <td>EEF1A1P5</td>\n",
       "      <td>61.478640</td>\n",
       "      <td>0.808213</td>\n",
       "      <td>0.808213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14852</th>\n",
       "      <td>EEF1G</td>\n",
       "      <td>45.863584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15024</th>\n",
       "      <td>EIF4B</td>\n",
       "      <td>22.274156</td>\n",
       "      <td>19.733755</td>\n",
       "      <td>19.733755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16782</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>346.996023</td>\n",
       "      <td>250.412455</td>\n",
       "      <td>250.412455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19979</th>\n",
       "      <td>ITGB1</td>\n",
       "      <td>22.378251</td>\n",
       "      <td>18.745036</td>\n",
       "      <td>18.745036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24934</th>\n",
       "      <td>MT-ATP8</td>\n",
       "      <td>86.415585</td>\n",
       "      <td>4.680054</td>\n",
       "      <td>4.680054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24935</th>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>50.031114</td>\n",
       "      <td>42.914260</td>\n",
       "      <td>42.914260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24943</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>2535.957351</td>\n",
       "      <td>13.684116</td>\n",
       "      <td>13.684116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24949</th>\n",
       "      <td>MT-TC</td>\n",
       "      <td>75.402070</td>\n",
       "      <td>0.759928</td>\n",
       "      <td>0.759928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25356</th>\n",
       "      <td>MYL6</td>\n",
       "      <td>20.234264</td>\n",
       "      <td>14.746390</td>\n",
       "      <td>14.746390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48341</th>\n",
       "      <td>RPL26</td>\n",
       "      <td>26.827455</td>\n",
       "      <td>5.184567</td>\n",
       "      <td>5.184567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48551</th>\n",
       "      <td>RPL41</td>\n",
       "      <td>21.093422</td>\n",
       "      <td>16.590253</td>\n",
       "      <td>16.590253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48764</th>\n",
       "      <td>RPS10</td>\n",
       "      <td>35.602033</td>\n",
       "      <td>0.277076</td>\n",
       "      <td>0.277076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49425</th>\n",
       "      <td>SCD</td>\n",
       "      <td>63.081806</td>\n",
       "      <td>34.998646</td>\n",
       "      <td>34.998646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49905</th>\n",
       "      <td>SHC1</td>\n",
       "      <td>28.804891</td>\n",
       "      <td>24.461643</td>\n",
       "      <td>24.461643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53424</th>\n",
       "      <td>TPT1</td>\n",
       "      <td>31.445395</td>\n",
       "      <td>17.710740</td>\n",
       "      <td>17.710740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54172</th>\n",
       "      <td>TUBA1B</td>\n",
       "      <td>30.612656</td>\n",
       "      <td>9.866877</td>\n",
       "      <td>9.866877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54185</th>\n",
       "      <td>TUBB</td>\n",
       "      <td>36.919341</td>\n",
       "      <td>26.592960</td>\n",
       "      <td>3.324120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54390</th>\n",
       "      <td>UBB</td>\n",
       "      <td>26.596593</td>\n",
       "      <td>9.118682</td>\n",
       "      <td>9.118682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54396</th>\n",
       "      <td>UBC</td>\n",
       "      <td>39.640373</td>\n",
       "      <td>16.111462</td>\n",
       "      <td>16.111462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54960</th>\n",
       "      <td>VIM</td>\n",
       "      <td>36.704718</td>\n",
       "      <td>19.648917</td>\n",
       "      <td>19.648917</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene    Stringtie  featureCount  featureCount_with0\n",
       "14799    EEF1A1   481.814875    401.810469          401.810469\n",
       "14833  EEF1A1P5    61.478640      0.808213            0.808213\n",
       "14852     EEF1G    45.863584      0.000000            0.000000\n",
       "15024     EIF4B    22.274156     19.733755           19.733755\n",
       "16782     GAPDH   346.996023    250.412455          250.412455\n",
       "19979     ITGB1    22.378251     18.745036           18.745036\n",
       "24934   MT-ATP8    86.415585      4.680054            4.680054\n",
       "24935    MT-CO1    50.031114     42.914260           42.914260\n",
       "24943   MT-ND4L  2535.957351     13.684116           13.684116\n",
       "24949     MT-TC    75.402070      0.759928            0.759928\n",
       "25356      MYL6    20.234264     14.746390           14.746390\n",
       "48341     RPL26    26.827455      5.184567            5.184567\n",
       "48551     RPL41    21.093422     16.590253           16.590253\n",
       "48764     RPS10    35.602033      0.277076            0.277076\n",
       "49425       SCD    63.081806     34.998646           34.998646\n",
       "49905      SHC1    28.804891     24.461643           24.461643\n",
       "53424      TPT1    31.445395     17.710740           17.710740\n",
       "54172    TUBA1B    30.612656      9.866877            9.866877\n",
       "54185      TUBB    36.919341     26.592960            3.324120\n",
       "54390       UBB    26.596593      9.118682            9.118682\n",
       "54396       UBC    39.640373     16.111462           16.111462\n",
       "54960       VIM    36.704718     19.648917           19.648917"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare.loc[abs(compare[\"featureCount\"]-compare[\"Stringtie\"])>(.1*compare[\"Stringtie\"])].loc[compare[\"Stringtie\"]>20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Stringtie</th>\n",
       "      <th>featureCount</th>\n",
       "      <th>featureCount_with0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7SK</td>\n",
       "      <td>0.011352</td>\n",
       "      <td>0.015343</td>\n",
       "      <td>0.001279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.042698</td>\n",
       "      <td>0.003610</td>\n",
       "      <td>0.003610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.044522</td>\n",
       "      <td>0.028881</td>\n",
       "      <td>0.028881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56631</th>\n",
       "      <td>ZYG11AP1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56632</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>1.137518</td>\n",
       "      <td>1.138087</td>\n",
       "      <td>1.138087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56633</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>6.098740</td>\n",
       "      <td>5.825361</td>\n",
       "      <td>5.825361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56634</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>1.384459</td>\n",
       "      <td>1.391245</td>\n",
       "      <td>1.391245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56635</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.216548</td>\n",
       "      <td>1.226986</td>\n",
       "      <td>1.226986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56636 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  Stringtie  featureCount  featureCount_with0\n",
       "0       5S_rRNA   0.000000      0.000000            0.000000\n",
       "1           7SK   0.011352      0.015343            0.001279\n",
       "2          A1BG   0.042698      0.003610            0.003610\n",
       "3      A1BG-AS1   0.044522      0.028881            0.028881\n",
       "4          A1CF   0.000000      0.000000            0.000000\n",
       "...         ...        ...           ...                 ...\n",
       "56631  ZYG11AP1   0.000000      0.000000            0.000000\n",
       "56632    ZYG11B   1.137518      1.138087            1.138087\n",
       "56633       ZYX   6.098740      5.825361            5.825361\n",
       "56634     ZZEF1   1.384459      1.391245            1.391245\n",
       "56635      ZZZ3   1.216548      1.226986            1.226986\n",
       "\n",
       "[56636 rows x 4 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#conclusion: featureCount and Stringtie are different, but will proceed with featureCount\n",
    "#significant (>10%) differences in 22 genes out of 56636"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "r-cpu.3-6.m65",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.3-6:m65"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
